Demo/FTP/client.cpp

```diff
@@ -0,0 +1,254 @@
+#include <netdb.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <fstream>
+#include <cstring>
+#include <cerrno>
+#include "../../Utils/Thread/ThreadPool.h"
+
+int sendAll(int sock, const std::string& msg) {
+    size_t total = 0;
+    size_t len = msg.size();
+    while (total < len) {
+        ssize_t sent = send(sock, msg.data() + total, len - total, 0);
+        if (sent <= 0) return 0;
+        total += sent;
+    }
+    return 1;
+}
+
+int readLine(int sock, std::string& out) {
+    out.clear();
+    char c;
+    while (recv(sock, &c, 1, 0) == 1) {
+        if (c == '\r') continue;
+        if (c == '\n') break;
+        out.push_back(c);
+    }
+    return out.empty() ? -1 : 1;
+}
+
+void client_handler(int sock) {
+
+}
+
+int main(int argc, char** argv) {
+    if (argc < 3) {
+        std::cerr << "Usage: " << argv[0] << " host port\n";
+        return 1;
+    }
+
+    const char* host = argv[1];
+    int port = std::atoi(argv[2]);
+
+    int ctrlSock = socket(AF_INET, SOCK_STREAM, 0);
+    if (ctrlSock < 0) {
+        std::cerr << "Error creating control socket: " << strerror(errno) << "\n";
+        return 1;
+    }
+
+    sockaddr_in servAddr{};
+    servAddr.sin_family = AF_INET;
+    servAddr.sin_port = htons(port);
+    if (inet_pton(AF_INET, host, &servAddr.sin_addr) != 1) {
+        std::cerr << "Invalid server IP address\n";
+        close(ctrlSock);
+        return 1;
+    }
+
+    if (connect(ctrlSock, (sockaddr*)&servAddr, sizeof(servAddr)) < 0) {
+        std::cerr << "Failed to connect control socket: " << strerror(errno) << "\n";
+        close(ctrlSock);
+        return 1;
+    }
+
+    std::string line;
+    ThreadPool threadPool;
+
+    if (readLine(ctrlSock, line) > 0)
+        std::cout << line << "\n";
+
+    int dataListenerSock = -1;
+    bool portSet = false;
+
+    std::string inputLine;
+    while (std::getline(std::cin, inputLine)) {
+        if (inputLine.empty()) continue;
+
+        std::istringstream iss(inputLine);
+        std::string cmd;
+        iss >> cmd;
+
+        if (cmd == "PORT") {
+            std::string params;
+            std::getline(iss, params);
+            params.erase(0, params.find_first_not_of(" \t"));
+
+            std::vector<int> nums;
+            std::istringstream pss(params);
+            std::string token;
+            while (std::getline(pss, token, ',')) {
+                try {
+                    nums.push_back(std::stoi(token));
+                } catch (...) {
+                    std::cerr << "Invalid PORT parameters.\n";
+                    nums.clear();
+                    break;
+                }
+            }
+            if (nums.size() != 6) {
+                std::cerr << "PORT command requires 6 comma-separated values.\n";
+                continue;
+            }
+
+            std::string ipStr = std::to_string(nums[0]) + "." + std::to_string(nums[1]) + "." +
```

Comment:
No input validation!
`nums[i] < 0 || nums[i] > 255` -> Invalid octet.

Reply:
Also this line builds many temporary strings. Consider using
`std::format("{}.{}.{}.{}", nums[0], nums[1], nums[2], nums[3]);`